README
NAME
    AnyMQ - Non-blocking message queue system based on AnyEvent

SYNOPSIS
      use AnyMQ;
      my $mq = AnyMQ->topic('Foo'); # gets an AnyMQ::Topic object
      $mq->publish({ message => 'Hello world'});

      #  bind to external message queue servers using traits.
      #  my $bus = AnyMQ->new_with_traits(traits => ['AMQP'],
      #                                   host   => 'localhost',
      #                                   port   => 5672,
      #                                   user   => 'guest',
      #                                   pass   => 'guest',
      #                                   vhost  => '/',
      #                                   exchange => '');
      #  my $mq = $bus->topic('foo')

      $mq->publish({ message => 'Hello world'});

      # $bus->new_listener('client_id', $mq);

DESCRIPTION
    AnyMQ is message queue system based on AnyEvent. It can store all
    messages in memory or use external message queue servers.

    Messages are published to AnyMQ::Topic, and consumed with AnyMQ::Queue.

METHODS
  new
    Returns a new AnyMQ object, which is a message bus that can associate
    with arbitrary AnyMQ::Topic and consumed by AnyMQ::Queue

  topic($name or %opt)
    Returns a AnyMQ::Topic with given name or constructor options %opt. If
    called as class method, the default bus will be used. Topics not known
    to the current AnyMQ bus will be created.

  new_topic($name or %opt)
    Creates and returns a new AnyMQ::Topic object with given name or
    constructor options %opt. This should not be called directly.

  new_listener(@topic)
    Returns a new AnyMQ::Queue object, and subscribes to the optional given
    topic. If called as class method, the default bus will be used.

AUTHORS
    Tatsuhiko Miyagawa Chia-liang Kao

LICENSE
    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

SEE ALSO
    AnyMQ::Topic, AnyMQ::Queue



Changes
Revision history for Perl extension AnyMQ

0.35 Sat Feb 16 18:34:24 CST 2013
        - Support unpoll. [alexmv]

0.34 Thu Feb  2 23:29:27 CST 2012
        - Improvements for traits to override publish method.

0.33 Sun Feb  6 21:33:30 CST 2011
        - Remove unused dependency.

0.32 Wed Dec 22 17:08:08 CST 2010
        - Use Any::Moose
        - Fix a bug when new messages are created during listener cb,
          poll might not be called immediately

0.31 Mon Mar 29 20:02:18 CST 2010
        - Topic and Queue recycling infrastructure

0.30 Wed Mar 24 22:16:14 CST 2010
        - Improve error and timeout handling
        - Better test coverage

0.29  Tue Mar  9 11:09:23 CST 2010
        - WithBacklog topic trait
        - Fix poll_once

0.28  Sun Mar  7 12:07:51 CST 2010
        - original version


t/topic-reaper.pl
use strict;
use Test::More;
use AnyMQ;
use AnyMQ::Topic;
my $bus = AnyMQ->new;

sub do_test {
    my $channel = $bus->topic({ name            => 'test',
                                reaper_interval => 0,
                                recycle         => 1 });

    my $cv = AE::cv;
    my $client = AnyMQ->new_listener($channel);
    $client->on_error( sub { $_[0]->destroyed(1); $cv->send });
    $client->poll(sub { die });

    $channel->publish({ data => 1});
    $channel->publish({ data => 2});

    $cv->recv;
    ok($client->destroyed, 'client is destroyed');
    ok($channel->has_no_listeners, 'listener destroyed');
}

do_test();

is_deeply([keys %{$bus->topics}], [], 'channel destroyed when all listeners are gone');

done_testing;


t/listener-reaper.pl
use strict;
use Test::More;
use AnyMQ;
use AnyMQ::Topic;
my $bus = AnyMQ->new;

my $channel = AnyMQ::Topic->new
     (bus => $bus, reaper_interval => 1);

my $cv = AE::cv;
my $client = AnyMQ->new_listener($channel);
$client->on_error( sub { $_[0]->destroyed(1); $cv->send });
$client->poll(sub { die });

$channel->publish({ data => 1});

$cv->recv;
ok($client->destroyed, 'client is destroyed');
ok(!$channel->has_no_listeners, 'listener still registered with topic');

my $cv = AE::cv;
my $w; $w = AnyEvent->timer(after => 2, cb => sub { $cv->send });

$cv->recv;

ok($channel->has_no_listeners, 'listener automatically reapped');

done_testing;


lib/AnyMQ/MQ.pm
package AnyMQ::MQ;;
use strict;

use AnyEvent;
use Any::Moose;
use Try::Tiny;
use Scalar::Util;
use Time::HiRes;
use constant DEBUG => 0;

has channel => (is => 'rw', isa => 'Str');
has backlog => (is => 'rw', isa => 'ArrayRef', default => sub { [] });
has clients => (is => 'rw', isa => 'HashRef', default => sub { +{} });

our $BacklogLength = 30; # TODO configurable

my %instances;

sub channels {
    values %instances;
}

sub instance {
    my($class, $name) = @_;
    $instances{$name} ||= $class->new(channel => $name);
}

sub backlog_events {
    my $self = shift;
    reverse grep defined, @{$self->backlog};
}

sub append_backlog {
    my($self, @events) = @_;
    my @new_backlog = (reverse(@events), @{$self->backlog});
    $self->backlog([ splice @new_backlog, 0, $BacklogLength ]);
}

sub publish {
    my($self, @events) = @_;

    for my $client_id (keys %{$self->clients}) {
        my $client = $self->clients->{$client_id};
        if ($client->{cv}->cb) {
            # currently listening: flush and send the events right away
            $self->flush_events($client_id, @events);
        } else {
            # between long poll comet: buffer the events
            # TODO: limit buffer length
            warn "Buffering new events for $client_id" if DEBUG;
            push @{$client->{buffer}}, @events;
        }
    }
    $self->append_backlog(@events);
}

sub flush_events {
    my($self, $client_id, @events) = @_;

    my $client = $self->clients->{$client_id} or return;
    try {
        my $cb = $client->{cv}->cb;
        $client->{cv}->send(@events);
        $client->{cv} = AE::cv;
        $client->{buffer} = [];

        if ($client->{persistent}) {
            $client->{cv}->cb($cb);
        } else {
            $client->{timer} = AE::timer 30, 0, sub {
                Scalar::Util::weaken $self;
                warn "Sweep $client_id (no long-poll reconnect)" if DEBUG;
                undef $client;
                delete $self->clients->{$client_id};
            };
            Scalar::Util::weaken $client->{timer};
        }
    } catch {
        warn $_;
    };
}

sub poll_once {
    my($self, $client_id, $cb, $timeout) = @_;

    my $is_new;
    my $client = $self->clients->{$client_id} ||= do {
        $is_new = 1;
        + { cv => AE::cv, persistent => 0, buffer => [] };
    };

    $client->{cv}->cb(sub { $cb->($_[0]->recv) });

    # reset garbage collection timeout with the long-poll timeout
    # $timeout = 0 is a valid timeout for interval-polling
    $timeout = 55 unless defined $timeout;
    $client->{timer} = AE::timer $timeout || 55, 0, sub {
        Scalar::Util::weaken $self;
        warn "Timing out $client_id long-poll" if DEBUG;
        $self->flush_events($client_id);
    };
    Scalar::Util::weaken $client->{timer};

    if ($is_new) {
        # flush backlog for a new client
        my @events = $self->backlog_events;
        $self->flush_events($client_id, @events) if @events;
    }elsif ( @{ $client->{buffer} } ) {
        # flush buffer for a long-poll client
        $self->flush_events($client_id, @{ $client->{buffer} });
    }
}

sub poll {
    my($self, $client_id, $cb) = @_;

    # TODO register client info like names and remote host in $client
    my $cv = AE::cv;
    $cv->cb(sub { $cb->($_[0]->recv) });
    my $s = $self->clients->{$client_id} = {
        cv => $cv, persistent => 1, buffer => [],
    };

    my @events = $self->backlog_events;
    $self->flush_events($client_id, @events) if @events;
}

1;

__END__

=encoding utf-8

=for stopwords

=head1 NAME

AnyMQ::MQ - AnyMQ Message Queue

=head1 SYNOPSIS

To publish a message, you first create an instance of the message queue on
a specific channel:

    my $mq = AnyMQ::MQ->instance($channel);
    $mq->publish({
        type => "message", data => $your_data,
        address => $self->request->address,
        time => scalar Time::HiRes::gettimeofday,
    });

Later, in a handler, you can poll for new messages:

    my $mq = AnyMQ::MQ->instance($channel);
    my $client_id = $self->request->param('client_id')
        or die;
    $mq->poll_once($client_id, sub { $self->write(\@_); $self->finish; });

Additionally, if you are using Multipart XmlHttpRequest (MXHR) you can use
the event API, and run a callback each time a new message is published:

    my $mq = AnyMQ::MQ->instance($channel);
    $mq->poll($client_id, sub {
        my @events = @_;
        for my $event (@events) {
            $self->stream_write($event);
        }
    });

=head1 DESCRIPTION

AnyMQ::MQ is a simple message queue, storing all messages in
memory, and keeping track of a configurable backlog.  All polling requests
are made with a C<$client_id>, and the message queue keeps track of a buffer
per client, to ensure proper message delivery.

=head1 CONFIGURATION

=over

=item BacklogLength

To configure the number of messages in the backlog, set 
C<$AnyMQ::MQ::BacklogLength>.  By default, this is set to 30.

=back

=head1 METHODS

=head2 publish

This method publishes a message into the message queue, for immediate 
consumption by all polling clients.

=head2 poll($client_id, $code_ref)

This is the event-driven poll mechanism, which accepts a callback as the
second parameter. It will stream messages to the code ref passed in. 

=head2 poll_once($client_id, $code_ref)

This method returns all messages since the last poll to the code reference
passed as the second parameter.

=head1 AUTHOR

Tatsuhiko Miyagawa

Chia-liang Kao

=head1 SEE ALSO

L<Tatsumaki>

=cut


lib/AnyMQ/Queue.pm
package AnyMQ::Queue;
use strict;
use Any::Moose;
use AnyEvent;
use Try::Tiny;
use Scalar::Util qw(weaken refaddr);
use Time::HiRes;
use constant DEBUG => 0;

has id => (is => 'rw', isa => 'Str');
has persistent => (is => "rw", isa => "Bool", default => sub { 0 });
has buffer => (is => "ro", isa => "ArrayRef", default => sub { [] });
has cv => (is => "rw", isa => "AnyEvent::CondVar", default => sub { AE::cv });
has destroyed => (is => "rw", isa => "Bool", default => sub { 0 });
has on_timeout => (is => "rw");
has on_error  => (is => "rw");
has timeout => (is => "rw", isa => "Int", default => sub { 55 });

sub BUILD {
    my $self = shift;
    $self->id('AnyMQ-'.refaddr($self)) unless $self->id;
}

sub append {
    my($self, @messages) = @_;

    if ($self->{cv}->cb) {
        # currently listening: flush and send the events right away
        $self->_flush(@messages);
    } else {
        # between long poll comet: buffer the events
        push @{$self->{buffer}}, @messages;
    }
}

sub subscribe {
    my ($self, $topic) = @_;
    Carp::cluck unless $topic;
    $topic->add_subscriber($self);
}

sub _flush {
    my ($self, @messages) = @_;

    my $cb = $self->{cv}->cb;
    my $cv = $self->{cv};
    $self->{cv} = AE::cv;
    $self->{buffer} = [];

    try {
        $cv->send(@messages);
    } catch {
        if ($self->on_error) {
            $self->on_error->($self, $_, @messages);
        }
        else {
            $self->destroyed(1);
        }
    };

    return if $self->destroyed;

    if ($self->{persistent}) {
        $self->{cv}->cb($cb);
        $self->_flush( @{ $self->{buffer} })
            if @{ $self->{buffer} };
    } else {
        $self->{timer} = $self->_reaper;
    }

}

sub _reaper {
    my ($self, $timeout) = @_;

    AnyEvent->timer(
        after => $timeout || $self->timeout,
        cb => sub {
            weaken $self;
            warn "Timing out $self long-poll" if DEBUG;
            if ($self->on_timeout) {
                $self->on_timeout->($self, "timeout")
            }
            else {
                $self->destroyed(1);
            }
        });
}

sub poll_once {
    my($self, $cb, $timeout) = @_;

    warn "already polled by another client" if $self->persistent;

    $self->{cv}->cb(sub { $cb->($_[0]->recv) });

    # reset garbage collection timeout with the long-poll timeout
    # $timeout = 0 is a valid timeout for interval-polling
    $timeout = 55 unless defined $timeout;
    $self->{timer} = AE::timer $timeout || 55, 0, sub {
        weaken $self;
        warn "long-poll timeout, flush and wait for client reconnect" if DEBUG;
        $self->_flush;
    };
    weaken $self->{timer};

    # flush buffer for a long-poll client
    $self->_flush( @{ $self->{buffer} })
        if @{ $self->{buffer} };
}

sub poll {
    my ($self, $cb) = @_;
    $self->cv->cb(sub { $cb->($_[0]->recv) });
    $self->persistent(1);

    undef $self->{timer};

    $self->_flush( @{ $self->{buffer} })
        if @{ $self->{buffer} };
}

sub unpoll {
    my $self = shift;
    $self->cv->cb(undef);
    $self->persistent(0);
    $self->{timer} = $self->_reaper;
}

__PACKAGE__->meta->make_immutable;
no Any::Moose;
1;

__END__

=encoding utf-8

=for stopwords

=head1 NAME

AnyMQ::Queue - AnyMQ Message Queue

=head1 SYNOPSIS

  my $channel = AnyMQ->topic('Foo');
  my $client = AnyMQ->new_listener($channel);

  $client->poll_once(sub {
      my @messages = @_;
      # ...
  });

  $client->poll(sub {
      my @messages = @_;
      # ...
  });

=head1 DESCRIPTION

An AnyMQ::Queue instance is a queue, each message put into the queue
can be consumed exactly once.  It's used as the client (or the
subscriber in terms of pub/sub) in L<AnyMQ>.  An AnyMQ::Queue can
subscribe to multiple L<AnyMQ::Topic>.

=head1 METHODS

=head2 subscribe($topic)

Subscribe to a L<AnyMQ::Topic> object.

=head2 poll($code_ref)

This is the event-driven poll mechanism, which accepts a callback.
Messages are streamed to C<$code_ref> passed in.

=head2 unpoll

Cancels a running L</poll>, which will result in L</on_timeout> being
called.

=head2 poll_once($code_ref, $timeout)

This method returns all messages since the last poll to C<$code_ref>.
It blocks for C<$timeout> seconds if there's currently no messages
available.

=head2 destroyed(BOOL)

Marking the current queue as destroyed or not.

=head2 timeout($seconds)

Timeout value for this queue.  Default is 55.

=head2 on_error(sub { my ($queue, $error, @msg) = @_; ... })

Sets the error handler invoked when C<poll> or C<poll_once> callbacks
fail.  By default the queue is marked as destroyed.  If you register
this error handler, you should call C<< $queue->destroyed(1) >> should you
wish to mark the queue as destroyed and reclaim resources.

Note that for queues that are currently C<poll>'ed, you may unset the
C<persistent> attribute to avoid the queue from being destroyed, and
can be used for further C<poll> or C<poll_once> calls.  In this case,
C<on_timeout> will be triggered if C<poll> or C<poll_once> is not
called after C<< $self->timeout >> seconds.

=head2 on_timeout(sub { my ($queue, $error) = @_; ... })

If a queue is not currently polled, this callback will be triggered
after C<< $self->timeout >> seconds.  The default behaviour is marking the
queue as destroyed.

=head2 append(@messages)

Append messages directly to the queue.  You probably want to use
C<publish> method of L<AnyMQ::Topic>

=head1 SEE ALSO

L<AnyMQ> L<AnyMQ::Topic>

=cut


lib/AnyMQ/Topic.pm
package AnyMQ::Topic;

use strict;
use 5.008_001;
our $VERSION = '0.01';

use AnyEvent;
use Any::Moose;
use Try::Tiny;
use Scalar::Util;
use Time::HiRes;

with any_moose("X::Traits");

has name => (is => 'rw', isa => 'Str');
has bus => (is => "ro", isa => "AnyMQ", weak_ref => 1);
has queues => (traits => ['Hash'],
               is => 'rw',
               isa => 'HashRef',
               default => sub { {} },
               handles => {
                   add_listener      => 'set',
                   has_no_listeners  => 'is_empty',
               }
           );
has recycle => (is => "rw", isa => "Bool", default => sub { 0 });
has 'reaper_interval' => (is => 'ro', isa => 'Int', default => sub { 30 });
has 'publish_to_queues' => (is => 'rw', isa => 'Bool', default => sub { 1 });
has '_listener_reaper' => (is => 'rw');
has '+_trait_namespace' => (default => 'AnyMQ::Topic::Trait');

sub BUILD {
    my $self = shift;
    $self->install_reaper if $self->reaper_interval;
}

sub install_reaper {
    my $self = shift;

    $self->_listener_reaper(
        AnyEvent->timer(interval => $self->reaper_interval,
                        cb => sub { $self->reap_destroyed_listeners })
    );
}

sub reap_destroyed_listeners {
    my $self = shift;
    return if $self->has_no_listeners;
    $self->remove_subscriber($_)
        for grep { $_->destroyed } values %{$self->queues};

    if ($self->recycle && $self->has_no_listeners) {
        delete $self->bus->topics->{$self->name};
    }
}

sub publish {
    my ($self, @messages) = @_;
    $self->append_to_queues(@messages) if $self->publish_to_queues;
    $self->dispatch_messages(@messages);
}

sub dispatch_messages {
    my ($self, @messages) = @_;
}

sub append_to_queues {
    my ($self, @messages) = @_;
    $self->reap_destroyed_listeners;
    for (values %{$self->queues}) {
        $_->append(@messages);
    }
    $self->install_reaper if $self->reaper_interval;
}

sub add_subscriber {
    my ($self, $queue) = @_;
    $self->add_listener($queue->id, $queue);
}

sub remove_subscriber {
    my ($self, $queue) = @_;
    delete $self->queues->{$queue->id};
}

__PACKAGE__->meta->make_immutable;
no Any::Moose;
1;

=encoding utf-8

=for stopwords

=head1 NAME

AnyMQ::Topic - AnyMQ Topic

=head1 SYNOPSIS

  my $channel = AnyMQ->topic('Foo');
  my $client = AnyMQ->new_listener($channel);

=head1 DESCRIPTION

An AnyMQ::Topic instance is a topic where messages can be published
to, and L<AnyMQ::Queue> objects can subscribe to.  each message
published to the topic will be appended to each subscribing queue.

=head1 ATTRIBUTES

=head2 recycle

True if the topic should be recycled once all listeners are gone.

=head2 reaper_interval

Interval in seconds that destroyed listeners to this topic should be
reaped and freed.

=head1 METHODS

=head2 publish(@messages)

Publish messages to the topic.

=head2 add_subscriber($queue)

Add a new listener to the topic.

=head1 SEE ALSO

L<AnyMQ> L<AnyMQ::Queue>

=cut


lib/AnyMQ/Topic/Trait/WithBacklog.pm
package AnyMQ::Topic::Trait::WithBacklog;
use strict;
use Any::Moose "::Role";

has backlog_length => (is => "rw", isa => "Int", default => sub { 30 });
has backlog => (is => 'rw', isa => 'ArrayRef', default => sub { [] });

sub backlog_events {
    my $self = shift;
    reverse grep defined, @{$self->backlog};
}

after 'publish' => sub {
    my ($self, @events) = @_;
    my @new_backlog = (reverse(@events), @{$self->backlog});
    $self->backlog([ splice @new_backlog, 0, $self->backlog_length ]);
};

after 'add_subscriber' => sub {
    my ($self, $queue) = @_;
    $queue->append($self->backlog_events);
};

1;


__END__

=encoding utf-8

=for stopwords

=head1 NAME

AnyMQ::Topic::Trait::WithBacklog - AnyMQ topic trait for backlog behaviour

=head1 SYNOPSIS

  my $bus = AnyMQ->new;
  my $channel = AnyMQ::Topic->new_with_trait
     (traits => ['WithBacklog'], backlog_length => 30, bus => $bus);
  my $client = AnyMQ->new_listener($channel);

=head1 DESCRIPTION

The topic trait for L<AnyMQ> provides backlog to a topic.  newly
subscribed listeners (an <AnyMQ::Queue> object) will received the last
C<backlog_length> messages published to the topic.

=head1 ATTRIBUTES

=head2 backlog_length

How many messages are to be kept for the topic.

=head1 SEE ALSO

L<AnyMQ> L<AnyMQ::Topic>

=cut




lib/AnyMQ.pm
package AnyMQ;
use strict;
use 5.008_001;
our $VERSION = '0.35';

use AnyEvent;
use Any::Moose;
use AnyMQ::Topic;
use AnyMQ::Queue;

with any_moose("X::Traits");

has '+_trait_namespace' => (default => 'AnyMQ::Trait');

has topics => (is => "ro", isa => "HashRef[AnyMQ::Topic]",
               default => sub { {} });

my $DEFAULT_BUS;

sub topic {
    my ($self, $opt) = @_;
    $opt = { name => $opt } unless ref $opt;
    $opt->{recycle} = 1 unless exists $opt->{recycle};

    unless (ref($self)) {
        $self = ($DEFAULT_BUS ||= $self->new);
    }

    $self->topics->{$opt->{name}} ||= $self->new_topic( $opt );
}

sub new_topic {
    my ($self, $opt) = @_;
    $opt = { name => $opt } unless ref $opt;
    AnyMQ::Topic->new( %$opt,
                       bus  => $self );
}

sub new_listener {
    my $self = shift;
    unless (ref($self)) {
        $self = ($DEFAULT_BUS ||= $self->new);
    }

    my $listener = AnyMQ::Queue->new;
    if (@_) {
        $listener->subscribe($_)
            for @_;
    }
    return $listener;
}

__PACKAGE__->meta->make_immutable;
no Any::Moose;
1;

__END__

=encoding utf-8

=for stopwords

=head1 NAME

AnyMQ - Non-blocking message queue system based on AnyEvent

=head1 SYNOPSIS

  use AnyMQ;
  my $mq = AnyMQ->topic('Foo'); # gets an AnyMQ::Topic object
  $mq->publish({ message => 'Hello world'});

  #  bind to external message queue servers using traits.
  #  my $bus = AnyMQ->new_with_traits(traits => ['AMQP'],
  #                                   host   => 'localhost',
  #                                   port   => 5672,
  #                                   user   => 'guest',
  #                                   pass   => 'guest',
  #                                   vhost  => '/',
  #                                   exchange => '');
  #  my $mq = $bus->topic('foo')

  $mq->publish({ message => 'Hello world'});

  # $bus->new_listener('client_id', $mq);

=head1 DESCRIPTION

AnyMQ is message queue system based on AnyEvent.  It can store all
messages in memory or use external message queue servers.

Messages are published to L<AnyMQ::Topic>, and consumed with
L<AnyMQ::Queue>.

=head1 METHODS

=head2 new

Returns a new L<AnyMQ> object, which is a message bus that can
associate with arbitrary L<AnyMQ::Topic> and consumed by
L<AnyMQ::Queue>

=head2 topic($name or %opt)

Returns a L<AnyMQ::Topic> with given name or constructor options
C<%opt>.  If called as class method, the default bus will be used.
Topics not known to the current AnyMQ bus will be created.

=head2 new_topic($name or %opt)

Creates and returns a new L<AnyMQ::Topic> object with given name or
constructor options C<%opt>.  This should not be called directly.

=head2 new_listener(@topic)

Returns a new L<AnyMQ::Queue> object, and subscribes to the optional
given topic.  If called as class method, the default bus will be used.

=head1 AUTHORS

Tatsuhiko Miyagawa
Chia-liang Kao

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 SEE ALSO

L<AnyMQ::Topic>, L<AnyMQ::Queue>

=cut
